File: src/lib/steering.ts
// Reusable "alive movement" core for ambient critters (cat) and NPCs (people). Three layers, per
// Craig Reynolds: action-selection (a tiny behaviour FSM) → steering (vehicle-model forces produce
// life-like paths) → locomotion (the components turn agent state into procedural transforms). Pure
// logic, no Svelte/three — components own the meshes. See docs/npc-movement.md.

const TAU = Math.PI * 2;
const rand = (a = 0, b = 1) => a + Math.random() * (b - a);

/**
 * Frame-rate-independent damped spring — the "secondary motion" workhorse (lean into turns, body
 * bob, tail follow-through, head look). A drop-in upgrade over lerp/tween: eases in, and (with
 * damping < 1) overshoots and settles like real mass. `step` integrates toward `target`.
 */
export class Spring {
	value: number;
	vel = 0;
	target: number;
	stiffness: number; // chase speed (higher = snappier)
	damping: number; // 1 = critical (no overshoot); < 1 = bouncy follow-through

	constructor(value = 0, stiffness = 9, damping = 0.85) {
		this.value = value;
		this.target = value;
		this.stiffness = stiffness;
		this.damping = damping;
	}

	step(dt: number, target = this.target): number {
		this.target = target;
		const k = this.stiffness;
		const c = 2 * Math.sqrt(k) * this.damping; // damping coeff from ratio
		const h = Math.min(dt, 0.05); // clamp for stability on frame spikes
		const a = -k * (this.value - this.target) - c * this.vel;
		this.vel += a * h;
		this.value += this.vel * h;
		return this.value;
	}
}

export type Behavior = 'wander' | 'pause' | 'lookAround' | 'sit' | 'groom' | 'pounce';

export interface AgentOpts {
	maxSpeed?: number;
	homeRadius?: number; // leash distance from home before steering back
	wanderRate?: number; // heading jitter (rad/s)
	accel?: number; // how hard it chases desired velocity (responsiveness)
	turnSpeed?: number; // how fast heading swings toward velocity
	wanderlust?: number; // chance (0..1) this agent is a far-roaming "explorer"
}

/**
 * An autonomous agent moving on the XZ plane. `update` runs the behaviour FSM, computes a steering
 * force (desired − current velocity), integrates, and turns the heading to follow velocity. The
 * component reads `speed`, `heading`, `turnRate`, `behavior` and `progress` to drive locomotion.
 */
export class Agent {
	x: number;
	z: number;
	vx = 0;
	vz = 0;
	heading: number; // facing angle; model nose is +Z, so group.rotation.y = heading
	speed = 0; // planar speed (m/s)
	turnRate = 0; // signed yaw rate (rad/s) → banking / tail lag

	hx: number; // home (leash centre)
	hz: number;
	wanderAngle: number;

	behavior: Behavior = 'wander';
	private elapsed = 0;
	private duration = 3;

	readonly maxSpeed: number; // public so the flock manager can scale forces to it
	private readonly homeRadius: number;
	private readonly wanderRate: number;
	private readonly accel: number;
	private readonly turnSpeed: number;
	readonly explorer: boolean; // a small fraction roam off to far-flung places
	personality: number; // 0..1 — varies cadence/speed so agents differ

	constructor(x: number, z: number, opts: AgentOpts = {}) {
		this.x = x;
		this.z = z;
		this.hx = x;
		this.hz = z;
		this.heading = rand(0, TAU);
		this.wanderAngle = this.heading;
		this.maxSpeed = opts.maxSpeed ?? 2.4;
		this.homeRadius = opts.homeRadius ?? 24;
		this.wanderRate = opts.wanderRate ?? 2.4;
		this.accel = opts.accel ?? 7;
		this.turnSpeed = opts.turnSpeed ?? 6;
		this.explorer = rand() < (opts.wanderlust ?? 0.14);
		this.personality = rand(0.3, 0.85);
		this.duration = rand(2, 5);
	}

	/** 0..1 progress through the current behaviour (drives pounce arcs, groom cycles, …). */
	get progress(): number {
		return Math.min(1, this.elapsed / this.duration);
	}

	/** Move the leash centre (e.g. keep a critter loosely near the player). */
	setHome(x: number, z: number): void {
		this.hx = x;
		this.hz = z;
	}

	private pick(menu: Behavior[]): void {
		// explorers occasionally strike out for a far-off place — relocate the home leash there, so
		// they actually journey across the map (over several wander cycles) and settle, then maybe
		// roam on again. Most agents are homebodies and never do this.
		if (this.explorer && Math.random() < 0.22) {
			const ang = rand(0, TAU);
			const far = rand(70, 200);
			this.hx = this.x + Math.sin(ang) * far;
			this.hz = this.z + Math.cos(ang) * far;
			this.behavior = 'wander';
			this.elapsed = 0;
			this.duration = rand(4, 8);
			return;
		}

		// weighted: heavily favour wander, then pauses, then the flavour behaviours
		const weights: Record<Behavior, number> = {
			wander: 5,
			pause: 2,
			lookAround: 1.4,
			groom: 1,
			sit: 1,
			pounce: 0.8
		};
		let total = 0;
		for (const b of menu) total += weights[b];
		let r = rand(0, total);
		let chosen: Behavior = menu[0];
		for (const b of menu) {
			r -= weights[b];
			if (r <= 0) {
				chosen = b;
				break;
			}
		}
		this.behavior = chosen;
		this.elapsed = 0;
		this.duration =
			chosen === 'wander'
				? rand(3, 7)
				: chosen === 'pounce'
					? rand(0.45, 0.7)
					: rand(1.6, 4.2);
	}

	update(dt: number, menu: Behavior[] = ['wander', 'pause'], flock?: { fx: number; fz: number }, boost = 1, forceMove = false): void {
		this.elapsed += dt;
		// while hunting/fleeing the manager forces continuous movement, so a chaser runs instead of
		// dropping into a sit/groom idle mid-pursuit (and its legs animate as running)
		if (forceMove) this.behavior = 'wander';
		else if (this.elapsed >= this.duration) this.pick(menu);

		// --- steering: desired velocity from the current behaviour ---
		let dvx = 0;
		let dvz = 0;
		let cap = this.maxSpeed;
		const moving = this.behavior === 'wander' || this.behavior === 'pounce';
		if (moving) {
			// Reynolds wander: a target on a circle ahead, nudged a small amount each frame
			this.wanderAngle += rand(-1, 1) * this.wanderRate * dt;
			const ahead = 1.5;
			const cx = this.x + Math.sin(this.heading) * ahead;
			const cz = this.z + Math.cos(this.heading) * ahead;
			let tx = cx + Math.sin(this.wanderAngle);
			let tz = cz + Math.cos(this.wanderAngle);
			// containment — Arrive back toward home when past the leash
			const homeDist = Math.hypot(this.x - this.hx, this.z - this.hz);
			if (homeDist > this.homeRadius) {
				tx = this.hx;
				tz = this.hz;
				this.wanderAngle = Math.atan2(this.hx - this.x, this.hz - this.z);
			}
			const tdx = tx - this.x;
			const tdz = tz - this.z;
			const td = Math.hypot(tdx, tdz) || 1;
			const burst = this.behavior === 'pounce' ? 2.3 : 1;
			cap = this.maxSpeed * burst;
			dvx = (tdx / td) * cap;
			dvz = (tdz / td) * cap;
		}
		// non-moving behaviours → desired 0 → Arrive smoothly to a stop (but flock can still nudge)

		// crowd flocking (separation/cohesion/alignment) is blended into the DESIRED velocity — not
		// straight into vx/vz — so it passes through the accel low-pass below and stays smooth.
		cap *= boost; // predator chase / prey panic-run speed multiplier (1 = normal)
		if (flock) {
			dvx += flock.fx;
			dvz += flock.fz;
			const dmag = Math.hypot(dvx, dvz);
			if (dmag > cap && dmag > 0) {
				dvx = (dvx / dmag) * cap;
				dvz = (dvz / dmag) * cap;
			}
		}

		// steering force = desired − current, applied as acceleration (the life-vs-mechanical fix)
		const k = Math.min(1, this.accel * dt);
		this.vx += (dvx - this.vx) * k;
		this.vz += (dvz - this.vz) * k;
		this.x += this.vx * dt;
		this.z += this.vz * dt;
		this.speed = Math.hypot(this.vx, this.vz);

		// --- heading follows velocity; record signed turn rate for banking / tail lag ---
		if (this.speed > 0.06) {
			const desired = Math.atan2(this.vx, this.vz);
			let dh = desired - this.heading;
			while (dh > Math.PI) dh -= TAU;
			while (dh < -Math.PI) dh += TAU;
			const turn = dh * Math.min(1, this.turnSpeed * dt);
			this.heading += turn;
			this.turnRate = turn / Math.max(dt, 1e-3);
		} else {
			this.turnRate *= Math.max(0, 1 - 4 * dt);
		}
	}

	/** Normalised gait drive (0..1) so locomotion swings faster the quicker the agent moves. */
	gaitRate(): number {
		return this.speed / this.maxSpeed;
	}
}

File: crates/worldsim/src/steering.rs
//! Agent steering — the deterministic port of `src/lib/steering.ts`'s `Agent` (action-selection FSM →
//! Reynolds-style wander → vehicle integration). This is the FIRST non-pure module: the live JS uses
//! `Math.random`, so there is no bit-parity with it — instead every random draw is now ADDRESSED by
//! `(seedId, tick, channel)` (the determinism migration the JS never did), making the sim reproducible and
//! thread-count-invariant (§6.8). The discrete state (behaviour picks + their TIMING) is bit-exact across
//! machines because it's driven only by the integer/bit-exact RNG and exact `elapsed = Σ dt`; only the
//! continuous positions differ in the last bits between platforms' `sin`/`cos`/`atan2` (≈1e-13), which is
//! immaterial. The parity test pins both: behaviour changes EXACTLY, positions to 1e-4, against a captured
//! JS reference that re-implements the same logic with this same addressed RNG. The `Spring` helper (lerp
//! secondary motion) is a VIEW concern and stays in JS.

use crate::simrng::{rand, range};

const TAU: f64 = std::f64::consts::TAU;
const PI: f64 = std::f64::consts::PI;

// RNG channels — one per draw SITE so two draws at the same (seedId, tick) don't correlate. Birth rolls key
// by (seedId, channel); per-tick rolls add the tick.
const CH_WANDER: i32 = 1;
const CH_PICK_RELOC: i32 = 2;
const CH_PICK_ANG: i32 = 3;
const CH_PICK_FAR: i32 = 4;
const CH_PICK_RELOC_DUR: i32 = 5;
const CH_PICK_CHOICE: i32 = 6;
const CH_PICK_CHOICE_DUR: i32 = 7;
const CH_HEADING: i32 = 10;
const CH_EXPLORER: i32 = 11;
const CH_PERSONALITY: i32 = 12;
const CH_DURATION: i32 = 13;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Behavior {
    Wander,
    Pause,
    LookAround,
    Groom,
    Sit,
    Pounce,
}

impl Behavior {
    fn weight(self) -> f64 {
        match self {
            Behavior::Wander => 5.0,
            Behavior::Pause => 2.0,
            Behavior::LookAround => 1.4,
            Behavior::Groom => 1.0,
            Behavior::Sit => 1.0,
            Behavior::Pounce => 0.8,
        }
    }
    #[inline]
    fn moving(self) -> bool {
        matches!(self, Behavior::Wander | Behavior::Pounce)
    }
    /// Stable u8 code for the JS render bridge (the renderers map it back to a pose). Order = enum order.
    pub fn code(self) -> u8 {
        match self {
            Behavior::Wander => 0,
            Behavior::Pause => 1,
            Behavior::LookAround => 2,
            Behavior::Groom => 3,
            Behavior::Sit => 4,
            Behavior::Pounce => 5,
        }
    }
}

pub struct AgentOpts {
    pub max_speed: f64,
    pub home_radius: f64,
    pub wander_rate: f64,
    pub accel: f64,
    pub turn_speed: f64,
    pub wanderlust: f64,
}

impl Default for AgentOpts {
    fn default() -> Self {
        Self { max_speed: 2.4, home_radius: 24.0, wander_rate: 2.4, accel: 7.0, turn_speed: 6.0, wanderlust: 0.14 }
    }
}

pub struct Agent {
    pub x: f64,
    pub z: f64,
    pub vx: f64,
    pub vz: f64,
    pub speed: f64,
    pub turn_rate: f64,
    pub heading: f64,
    pub wander_angle: f64,
    pub hx: f64, // home (leash centre)
    pub hz: f64,
    pub behavior: Behavior,
    pub elapsed: f64,
    pub duration: f64,
    pub seed_id: i32,
    pub explorer: bool,
    pub personality: f64,
    pub max_speed: f64, // public so the flock manager can scale forces to it
    home_radius: f64,
    wander_rate: f64,
    accel: f64,
    turn_speed: f64,
}

impl Agent {
    pub fn new(x: f64, z: f64, seed_id: i32, o: &AgentOpts) -> Self {
        let heading = range(0.0, TAU, &[seed_id, CH_HEADING]);
        Agent {
            x,
            z,
            vx: 0.0,
            vz: 0.0,
            speed: 0.0,
            turn_rate: 0.0,
            heading,
            wander_angle: heading,
            hx: x,
            hz: z,
            behavior: Behavior::Wander,
            elapsed: 0.0,
            duration: range(2.0, 5.0, &[seed_id, CH_DURATION]),
            seed_id,
            explorer: rand(&[seed_id, CH_EXPLORER]) < o.wanderlust,
            personality: range(0.3, 0.85, &[seed_id, CH_PERSONALITY]),
            max_speed: o.max_speed,
            home_radius: o.home_radius,
            wander_rate: o.wander_rate,
            accel: o.accel,
            turn_speed: o.turn_speed,
        }
    }

    /// Move the leash centre (e.g. keep a critter loosely near the player).
    pub fn set_home(&mut self, x: f64, z: f64) {
        self.hx = x;
        self.hz = z;
    }

    fn pick(&mut self, tick: i32, menu: &[Behavior]) {
        let s = self.seed_id;
        // explorers occasionally strike out for a far-off place — relocate the leash there.
        if self.explorer && rand(&[s, tick, CH_PICK_RELOC]) < 0.22 {
            let ang = range(0.0, TAU, &[s, tick, CH_PICK_ANG]);
            let far = range(70.0, 200.0, &[s, tick, CH_PICK_FAR]);
            self.hx = self.x + ang.sin() * far;
            self.hz = self.z + ang.cos() * far;
            self.behavior = Behavior::Wander;
            self.elapsed = 0.0;
            self.duration = range(4.0, 8.0, &[s, tick, CH_PICK_RELOC_DUR]);
            return;
        }
        // weighted: heavily favour wander, then pauses, then the flavour behaviours.
        let total: f64 = menu.iter().map(|b| b.weight()).sum();
        let mut r = range(0.0, total, &[s, tick, CH_PICK_CHOICE]);
        let mut chosen = menu[0];
        for &b in menu {
            r -= b.weight();
            if r <= 0.0 {
                chosen = b;
                break;
            }
        }
        self.behavior = chosen;
        self.elapsed = 0.0;
        self.duration = match chosen {
            Behavior::Wander => range(3.0, 7.0, &[s, tick, CH_PICK_CHOICE_DUR]),
            Behavior::Pounce => range(0.45, 0.7, &[s, tick, CH_PICK_CHOICE_DUR]),
            _ => range(1.6, 4.2, &[s, tick, CH_PICK_CHOICE_DUR]),
        };
    }

    /// One sim step. `flock` is the manager's Reynolds force (None in isolation); `boost` is the chase/flee
    /// speed multiplier; `force_move` keeps a chaser running instead of idling.
    pub fn update(&mut self, tick: i32, dt: f64, menu: &[Behavior], flock: Option<(f64, f64)>, boost: f64, force_move: bool) {
        self.elapsed += dt;
        if force_move {
            self.behavior = Behavior::Wander;
        } else if self.elapsed >= self.duration {
            self.pick(tick, menu);
        }

        let mut dvx = 0.0;
        let mut dvz = 0.0;
        let mut cap = self.max_speed;
        if self.behavior.moving() {
            // Reynolds wander: a target on a circle ahead, nudged a little each tick.
            self.wander_angle += range(-1.0, 1.0, &[self.seed_id, tick, CH_WANDER]) * self.wander_rate * dt;
            let cx = self.x + self.heading.sin() * 1.5;
            let cz = self.z + self.heading.cos() * 1.5;
            let mut tx = cx + self.wander_angle.sin();
            let mut tz = cz + self.wander_angle.cos();
            // containment — Arrive back toward home when past the leash.
            let home_dist = (self.x - self.hx).hypot(self.z - self.hz);
            if home_dist > self.home_radius {
                tx = self.hx;
                tz = self.hz;
                self.wander_angle = (self.hx - self.x).atan2(self.hz - self.z);
            }
            let tdx = tx - self.x;
            let tdz = tz - self.z;
            let td = {
                let h = tdx.hypot(tdz);
                if h == 0.0 {
                    1.0
                } else {
                    h
                }
            };
            let burst = if matches!(self.behavior, Behavior::Pounce) { 2.3 } else { 1.0 };
            cap = self.max_speed * burst;
            dvx = (tdx / td) * cap;
            dvz = (tdz / td) * cap;
        }

        cap *= boost;
        if let Some((fx, fz)) = flock {
            dvx += fx;
            dvz += fz;
            let dmag = dvx.hypot(dvz);
            if dmag > cap && dmag > 0.0 {
                dvx = (dvx / dmag) * cap;
                dvz = (dvz / dmag) * cap;
            }
        }

        // steering force = desired − current, applied as acceleration (the life-vs-mechanical fix).
        let k = (self.accel * dt).min(1.0);
        self.vx += (dvx - self.vx) * k;
        self.vz += (dvz - self.vz) * k;
        self.x += self.vx * dt;
        self.z += self.vz * dt;
        self.speed = self.vx.hypot(self.vz);

        // heading follows velocity; record signed turn rate for banking / tail lag.
        if self.speed > 0.06 {
            let desired = self.vx.atan2(self.vz);
            let mut dh = desired - self.heading;
            while dh > PI {
                dh -= TAU;
            }
            while dh < -PI {
                dh += TAU;
            }
            let turn = dh * (self.turn_speed * dt).min(1.0);
            self.heading += turn;
            self.turn_rate = turn / dt.max(1e-3);
        } else {
            self.turn_rate *= (1.0 - 4.0 * dt).max(0.0);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    const DT: f64 = 1.0 / 30.0;

    fn close(got: f64, want: f64) {
        assert!((got - want).abs() < 1e-4, "expected ~{want}, got {got}");
    }

    // Reference captured from src/lib/steering.ts's logic re-run with this addressed RNG (the determinism
    // migration). Behaviour changes + their TIMING are bit-exact (rng/integer driven); positions match to
    // 1e-4 (platform sin/cos/atan2 last-bits + the reference's 6-dp rounding).
    #[test]
    fn steering_parity() {
        let menu = [Behavior::Wander, Behavior::Pause, Behavior::LookAround, Behavior::Groom, Behavior::Sit];
        let o = AgentOpts { max_speed: 3.0, home_radius: 40.0, wander_rate: 1.3, accel: 7.0, turn_speed: 5.0, wanderlust: 0.55 };
        let mut a = Agent::new(0.0, 0.0, 12345, &o);

        // birth rolls
        close(a.heading, 5.11553);
        assert!(a.explorer);
        close(a.personality, 0.808471);
        close(a.duration, 3.191605);

        let mut changes: Vec<(i32, Behavior)> = Vec::new();
        let mut last = a.behavior;
        for t in 1..=250 {
            a.update(t, DT, &menu, None, 1.0, false);
            if a.behavior != last {
                changes.push((t, a.behavior));
                last = a.behavior;
            }
            match t {
                1 => {
                    close(a.x, -0.021331);
                    close(a.z, 0.009457);
                    close(a.heading, 5.117892);
                    assert_eq!(a.behavior, Behavior::Wander);
                }
                50 => {
                    close(a.x, -3.930816);
                    close(a.z, 2.507159);
                }
                120 => {
                    close(a.x, -7.778401);
                    close(a.z, 5.421086);
                    assert_eq!(a.behavior, Behavior::Pause);
                    close(a.duration, 3.735077);
                }
                250 => {
                    close(a.x, -10.75285);
                    close(a.z, 7.898893);
                    assert_eq!(a.behavior, Behavior::Wander);
                    close(a.duration, 4.821398);
                }
                _ => {}
            }
        }
        // bit-exact discrete events: pause at tick 96, back to wander at 209
        assert_eq!(changes, vec![(96, Behavior::Pause), (209, Behavior::Wander)]);
    }

    #[test]
    fn reproducible() {
        let menu = [Behavior::Wander, Behavior::Pause];
        let o = AgentOpts::default();
        let run = || {
            let mut a = Agent::new(3.0, -2.0, 999, &o);
            for t in 1..=300 {
                a.update(t, DT, &menu, None, 1.0, false);
            }
            (a.x, a.z, a.heading, a.behavior)
        };
        let r1 = run();
        let r2 = run();
        assert_eq!(r1.0.to_bits(), r2.0.to_bits()); // bit-identical → deterministic
        assert_eq!(r1.1.to_bits(), r2.1.to_bits());
        assert_eq!(r1.2.to_bits(), r2.2.to_bits());
        assert_eq!(r1.3, r2.3);
    }
}
